File: macos/jules/SessionPollingController.swift
import Foundation
import Combine

@MainActor
class SessionPollingController {
    private let sessionRepository: SessionRepository
    private weak var dataManager: DataManager?
    private var cancellables: Set<AnyCancellable> = []

    // MARK: - Unified Polling Timer

    // Single timer with 5-second base interval
    // Different operations run on different tick cycles:
    // - Active sessions: every 2 ticks (10s)
    // - Backfill: every 6 ticks (30s)
    // - Completed sessions: every 12 ticks (60s)
    // - Stale session check: every 120 ticks (10 minutes)
    private static let baseInterval: TimeInterval = 5
    private static let activeSessionsTicks = 2      // 10s
    private static let backfillTicks = 6            // 30s
    private static let completedSessionsTicks = 12  // 60s
    private static let staleSessionCheckTicks = 120 // 10 minutes
    private static let initialBackfillTicks = 2     // 10s delay before first backfill

    nonisolated(unsafe) private var unifiedTimer: Timer?
    private var tickCount: Int = 0
    private var isPaused: Bool = false
    private var hasRunInitialBackfill: Bool = false

    init(sessionRepository: SessionRepository, dataManager: DataManager) {
        self.sessionRepository = sessionRepository
        self.dataManager = dataManager
    }

    func startPolling() {
        stopPolling()
        tickCount = 0
        hasRunInitialBackfill = false
        isPaused = false

        // Single unified timer fires every 5 seconds
        unifiedTimer = Timer.scheduledTimer(withTimeInterval: Self.baseInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                await self?.onTick()
            }
        }
    }

    private func onTick() async {
        guard !isPaused else { return }

        tickCount += 1

        // Capture the current tick value for this invocation.
        // This is critical because multiple onTick() tasks can run concurrently
        // (timer fires every 5s regardless of whether previous tick completed).
        // Without capturing, async operations that take longer than 5s would see
        // a stale tickCount after resuming, causing missed or duplicate polls.
        let currentTick = tickCount

        // Initial backfill after 10 seconds (2 ticks)
        if !hasRunInitialBackfill && currentTick >= Self.initialBackfillTicks {
            hasRunInitialBackfill = true
            await backfillActivities()
        }

        // Active sessions: every 3 ticks (15s)
        if currentTick % Self.activeSessionsTicks == 0 {
            await pollActiveSessions()
        }

        // Backfill: every 6 ticks (30s)
        if currentTick % Self.backfillTicks == 0 {
            await backfillActivities()
        }

        // Completed sessions: every 12 ticks (60s)
        if currentTick % Self.completedSessionsTicks == 0 {
            await pollCompletedSessions()
        }

        // Stale session check: every 120 ticks (10 minutes)
        // Marks active sessions that haven't had updates for over an hour as completedUnknown
        if currentTick % Self.staleSessionCheckTicks == 0 {
            await checkStaleSessions()
        }
    }

    nonisolated func stopPolling() {
        unifiedTimer?.invalidate()
        unifiedTimer = nil
    }

    /// Pause all polling operations (e.g., during heavy UI operations)
    func pausePolling() {
        isPaused = true
    }

    /// Resume polling operations
    func resumePolling() {
        isPaused = false
    }

    private func pollActiveSessions() async {
        guard let dataManager = dataManager else { return }
        await sessionRepository.refresh()

        // Start with the top 5 active sessions
        let topActiveSessions = dataManager.sessions.prefix(5).filter { !$0.state.isTerminal }
        var idsToFetch = Set(topActiveSessions.map { $0.id })

        // Always include the currently active session window, if it exists and is not already in the list
        if let activeId = dataManager.activeSessionId, !idsToFetch.contains(activeId) {
            idsToFetch.insert(activeId)
        }

        if !idsToFetch.isEmpty {
            await dataManager.fetchActivities(for: Array(idsToFetch))
        }
    }

    private func pollCompletedSessions() async {
        guard let dataManager = dataManager else { return }
        await sessionRepository.refresh()
        // Only poll completed/completedUnknown sessions that still need stats (don't have cached git stats yet)
        let completedSessions = dataManager.sessions.prefix(5).filter {
            ($0.state == .completed || $0.state == .completedUnknown) && $0.needsActivityFetchForStats
        }
        let idsToFetch = completedSessions.map { $0.id }
        if !idsToFetch.isEmpty {
            await dataManager.fetchActivities(for: idsToFetch)
        }
    }

    private func backfillActivities() async {
        guard let dataManager = dataManager else { return }

        // Query the database directly for sessions that need backfilling.
        // This bypasses the in-memory limit (default 25) to find older sessions
        // that haven't had their activities fetched or git stats computed.
        // Skip the first 5 sessions (handled by pollActiveSessions/pollCompletedSessions)
        // and process up to 3 at a time for better performance.
        let sessionsToBackfill = await sessionRepository.getSessionsNeedingBackfill(offset: 5, limit: 3)

        let idsToBackfill = sessionsToBackfill.map { $0.id }
        if !idsToBackfill.isEmpty {
            await dataManager.fetchActivities(for: idsToBackfill)
        }
    }

    private func checkStaleSessions() async {
        // Check for sessions that have been active for too long without updates.
        // For each stale session, verifies with the API:
        // - If session was deleted (404), removes it from local DB
        // - If session state changed on server, updates local state
        let processedCount = await sessionRepository.markStaleSessions()
        if processedCount > 0 {
            print("📋 Processed \(processedCount) stale session(s)")
        }
    }
}

File: src/ui/session_polling_controller.cpp
#include "ui/session_polling_controller.h"

#include <QDebug>

namespace jules {

SessionPollingController::SessionPollingController(SessionRepository* repository,
                                                   JulesApiClient* apiClient,
                                                   QObject* parent)
    : QObject(parent)
    , m_repository(repository)
    , m_apiClient(apiClient)
    , m_timer(new QTimer(this))
{
    connect(m_timer, &QTimer::timeout, this, &SessionPollingController::onTick);
    connect(m_apiClient, &JulesApiClient::activitiesReceived,
            this, &SessionPollingController::onActivitiesReceived);
}

SessionPollingController::~SessionPollingController() {
    stopPolling();
}

void SessionPollingController::startPolling() {
    stopPolling();
    m_tickCount = 0;
    m_hasRunInitialBackfill = false;
    m_isPaused = false;
    
    m_timer->start(BASE_INTERVAL_MS);
}

void SessionPollingController::stopPolling() {
    m_timer->stop();
}

bool SessionPollingController::isPolling() const {
    return m_timer->isActive();
}

void SessionPollingController::pausePolling() {
    m_isPaused = true;
}

void SessionPollingController::resumePolling() {
    m_isPaused = false;
}

bool SessionPollingController::isPaused() const {
    return m_isPaused;
}

void SessionPollingController::setActiveSessionId(const QString& sessionId) {
    m_activeSessionId = sessionId;
}

QString SessionPollingController::activeSessionId() const {
    return m_activeSessionId;
}

void SessionPollingController::onTick() {
    if (m_isPaused) return;
    
    m_tickCount++;
    int currentTick = m_tickCount;
    
    // Initial backfill after 10 seconds (2 ticks)
    if (!m_hasRunInitialBackfill && currentTick >= INITIAL_BACKFILL_TICKS) {
        m_hasRunInitialBackfill = true;
        backfillActivities();
    }
    
    // Active sessions: every 2 ticks (10s)
    if (currentTick % ACTIVE_SESSIONS_TICKS == 0) {
        pollActiveSessions();
    }
    
    // Backfill: every 6 ticks (30s)
    if (currentTick % BACKFILL_TICKS == 0) {
        backfillActivities();
    }
    
    // Completed sessions: every 12 ticks (60s)
    if (currentTick % COMPLETED_SESSIONS_TICKS == 0) {
        pollCompletedSessions();
    }
    
    // Stale session check: every 120 ticks (10 minutes)
    if (currentTick % STALE_SESSION_CHECK_TICKS == 0) {
        checkStaleSessions();
    }
    
    emit tickCompleted(currentTick);
}

void SessionPollingController::pollActiveSessions() {
    emit requestSessionsRefresh();
    
    // Get top 5 active sessions
    QList<Session> allSessions = m_repository->getAllSessions();
    QStringList idsToFetch;
    
    int count = 0;
    for (const auto& session : allSessions) {
        if (count >= 5) break;
        if (session.isActive()) {
            idsToFetch.append(session.id);
            count++;
        }
    }
    
    // Always include the currently active session if not already in list
    if (!m_activeSessionId.isEmpty() && !idsToFetch.contains(m_activeSessionId)) {
        idsToFetch.append(m_activeSessionId);
    }
    
    if (!idsToFetch.isEmpty()) {
        for (const QString& id : idsToFetch) {
            if (!m_pendingActivityFetches.contains(id)) {
                m_pendingActivityFetches.insert(id);
                m_apiClient->getActivities(id);
            }
        }
        emit requestActivitiesFetch(idsToFetch);
    }
}

void SessionPollingController::pollCompletedSessions() {
    emit requestSessionsRefresh();
    
    // Only poll completed sessions that still need stats
    QList<Session> allSessions = m_repository->getAllSessions();
    QStringList idsToFetch;
    
    int count = 0;
    for (const auto& session : allSessions) {
        if (count >= 5) break;
        if ((session.state == SessionState::Completed || 
             session.state == SessionState::CompletedUnknown) &&
            session.needsActivityFetchForStats()) {
            idsToFetch.append(session.id);
            count++;
        }
    }
    
    if (!idsToFetch.isEmpty()) {
        for (const QString& id : idsToFetch) {
            if (!m_pendingActivityFetches.contains(id)) {
                m_pendingActivityFetches.insert(id);
                m_apiClient->getActivities(id);
            }
        }
        emit requestActivitiesFetch(idsToFetch);
    }
}

void SessionPollingController::backfillActivities() {
    // Query sessions that need backfilling (skip first 5, get next 3)
    QList<Session> allSessions = m_repository->getAllSessions();
    QStringList idsToBackfill;
    
    int skipped = 0;
    int count = 0;
    for (const auto& session : allSessions) {
        if (skipped < 5) {
            skipped++;
            continue;
        }
        if (count >= 3) break;
        
        // Backfill sessions that need activity fetching for stats
        if (session.needsActivityFetchForStats()) {
            idsToBackfill.append(session.id);
            count++;
        }
    }
    
    if (!idsToBackfill.isEmpty()) {
        for (const QString& id : idsToBackfill) {
            if (!m_pendingActivityFetches.contains(id)) {
                m_pendingActivityFetches.insert(id);
                m_apiClient->getActivities(id);
            }
        }
        emit requestActivitiesFetch(idsToBackfill);
    }
}

void SessionPollingController::checkStaleSessions() {
    // Check for sessions that have been active for too long without updates
    QList<Session> allSessions = m_repository->getAllSessions();
    
    for (const auto& session : allSessions) {
        if (session.isStaleActive()) {
            emit requestStaleSessionCheck(session.id);
        }
    }
}

void SessionPollingController::onActivitiesReceived(const QString& sessionId,
                                                     const QList<Activity>& activities) {
    m_pendingActivityFetches.remove(sessionId);
    
    // Update session with activities and compute git stats
    auto sessionOpt = m_repository->getSession(sessionId);
    if (sessionOpt.has_value()) {
        Session session = sessionOpt.value();
        session.activities = activities;
        session.updateCachedDiffData();
        m_repository->saveSession(session);
    }
}

} // namespace jules
